INT-4219: Add @UseSpelInvoker

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4219

Needs (a lot) more work, tests, docs.

Just want to get buy-in for the approach.

Also see note about resolving placeholders for compiler mode.

garyrussell committed:
INT-4219: Add @UseSpelInvoker

JIRA: https://jira.spring.io/browse/INT-4219

artembilan reviewed (changes requested):
I we're good to go with this approach and that doesn't look so ugly if you don't like to use `InvocableMessageHandler`.

I don't see how you apply that `compilerMode`, but I guess that willl in upcoming commits.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -243,6 +243,10 @@ protected String getInputChannelAttribute() {
 	}
 
+	protected ConfigurableListableBeanFactory getBeanFactory() {
+		return this.beanFactory;
```
Yeah - this is some left-over dross from when I was trying to use the BF too early.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java:
```diff
@@ -0,0 +1,71 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+@Documented
+public @interface UseSpelInvoker {
```
I don't think so; we will need that for XML config (either a new attribute or they can autowire the handler and set the flag).

artembilan replied:
??? Even for XML config the POJO method can be marked with the `@UseSpelInvoker`.
We don't prohibit to use `@ServiceActivator` and similar when we parse XML config.
Finally everything ends up in the `MessagingMethodInvokerHelper` which just reads a `Method` for particular options and generates SpEL and `InvocableHandlerMethod` for runtime.

garyrussell replied:
The problem is they might want to invoke a class/method they don't own and can't annotate.

garyrussell committed:
Defer Expression Compilation; Resolve Placeholders

garyrussell committed:
Uncomment BeanExpressionResolver Stuff

garyrussell commented:
Next iteration.

artembilan commented:
Well, I think the progress is good here.
You can go ahead what you want to do here else.

And I see your point about an external class to use from XML.
There is always a workaround like local wrapper around that remote method.
My point just don't spend time for XML effort.
At least until strong requirements.

garyrussell commented:
>My point just don't spend time for XML effort.

Agreed.

artembilan commented:
FYI: there are some failures according Travis report.

garyrussell committed:
Fix Test Failures (Deferred Expression Parsing)

garyrussell committed:
More Test Fixes

Broke tests that manipulated the EXPRESSION_PARSER - change
EXPRESSION_PARSER_DEFAULT to reference EXPRESSION_PARSER from the super class.

artembilan commented:
Well, looks like we need only docs on the matter already!
And resolving conflicts after rebase.
Do you have anything else in mind to do for this task ?

garyrussell commented:
PR soon (tests and docs) - doing a `testAll` right now 😄

garyrussell committed:
More Tests and Docs

garyrussell committed:
Doc Polishing

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java:
```diff
@@ -0,0 +1,74 @@
+	 * invoked using SpEL instead of an
+	 * {@link org.springframework.messaging.handler.invocation.InvocableHandlerMethod}
+	 * with the specified compilerMode. If left empty, the default runtime compiler
+	 * mode will be used. Must evaluate to a String containing a valide compiler mode.
```
Typo: valid

Will fix on merge

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java:
```diff
@@ -0,0 +1,74 @@
+	 * invoked using SpEL instead of an
+	 * {@link org.springframework.messaging.handler.invocation.InvocableHandlerMethod}
+	 * with the specified compilerMode. If left empty, the default runtime compiler
+	 * mode will be used. Must evaluate to a String containing a valide compiler mode.
```
DITO.
Copy/paste artefact 😄

artembilan reviewed (approved):
A couple (in fact one) typos.
Merging...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/a50d866f81888f27c56f7298d946b786c52e9281
